### PR TITLE
improve realloc performance

### PR DESCRIPTION
**Pull request by adrian-gierakowski**
avoid unnecessary copying when reallocating command buffer

this is a proper implementation of what was intended in 92b47664b20201eb95dc1526b83a97ae13179831

and a few other fixes, see individual commits for details

I am going to follow up with a few other commits shortly

### CODE REVIEW

**Comment by adrian-gierakowski** [comment]
the commits from ba8d73771958425cbcb21adee43d81e778bfb727 onward are about reducing number of reallocations 
see travis log for how many reallocations are needed with the current buffer growth strategy: https://travis-ci.org/h0x91b/redis-fast-driver/jobs/360343899

5bb94241222a8104fd0b20b74faee384267aa142 this is reduced to 1

File: src/redis-fast-driver.cc [hunk 1]
```diff
@@ -276,17 +276,41 @@ NAN_METHOD(RedisConnector::RedisCmd) {
...
 			// LOG("bufsize is not big enough, current: %zu ", bufsize);
 			// bufsize = bufsize * 2;
-			bufsize = ((bufused + len) / 256 + 1) * 256;
+			// bufsize = ((bufused + len) / 256 + 1) * 256;
+			if (i == arraylen - 1) {
+				// last element
+				// only (bufused + len) is really needed but give it a bit of headroom 
+				// since its likely a similar command will be fired again (and it might
+				// be slightly larger)
+				bufsize = 1.2 * (bufused + len);
```
Comment (h0x91b) [inline src/redis-fast-driver.cc]: Nope, 1.2 multiplier is a bad strategy, for the best performance we should allocate memory aligned blocks. (Align is 64, so my 256 is aligned too, but in your case 4096*1.2 = 4915 which is not good)

Example from StackOverflow:
```
// cache line
#define ALIGN 64

void *aligned_malloc(int size) {
    void *mem = malloc(size+ALIGN+sizeof(void*));
    void **ptr = (void**)((uintptr_t)(mem+ALIGN+sizeof(void*)) & ~(ALIGN-1));
    ptr[-1] = mem;
    return ptr;
}

void aligned_free(void *ptr) {
    free(((void**)ptr)[-1]);
}
```

File: src/redis-fast-driver.cc [hunk 2]
```diff
@@ -276,17 +276,41 @@ NAN_METHOD(RedisConnector::RedisCmd) {
...
+			} else {
+				// estimate remaining space that is needed
+				float avarage_arg_size = float(bufused + len) / i;
+				// this is definitely needed
+				bufsize = bufused + len +
+					// estimated bytes needed for the remainder of the args
+					(arraylen - 1 - i) * avarage_arg_size
+					// plus some extra headroom
+					* 1.2;
```
Comment (h0x91b) [inline src/redis-fast-driver.cc]: Same here should be aligned by 64 bytes.

File: package-lock.json [hunk 3]
```diff
@@ -1,13 +1,1865 @@
 {
   "name": "redis-fast-driver",
-  "version": "2.0.2",
+  "version": "2.1.0",
   "lockfileVersion": 1,
   "requires": true,
   "dependencies": {
```
Comment (h0x91b) [inline package-lock.json]: Package-lock should be generated only for prod dependencies, please run:

`NODE_ENV=production npm install`

**Comment by h0x91b** [comment]
1.2 strategy must be aligned by 64 bytes for a performance. And several small things, please see review.

**Comment by adrian-gierakowski** [comment]
@h0x91b I've done what you asked for, although I am not sure if rounding up to multiple of 64 actually improves anythings.

1. if the purpose is to have memory page aligned the we should also adjust the pointer to point to start of the page, which we are not doing
2. as this article suggests https://stackoverflow.com/a/3190185, it might be better to simply tell malloc exactly how much memory you need and let it figure out how to efficiently allocate it.

anyway, I guess no harm rounding it :)

**Comment by adrian-gierakowski** [comment]
@h0x91b, I believe I have implemented everything you've requested. Do you have any further doubts or comments?

**Pull request merged**